File: packages/restapi/src/lib/utils/parseAPI.ts
import { ApiNotificationType, ParsedResponseType } from '../types';
import { decryptFeed } from './decryptFeed';

/**
 * @description parse the response gotten from the API
 * @param {ApiNotificationType[]} response
 * @returns {ParsedResponseType[]}
 */

/**
 * @description parse the response gotten from the API
 * @param {ApiNotificationType[]} response
 * @returns {Promise<ParsedResponseType[]>}
 */
export async function parseApiResponse(response: ApiNotificationType[], pgpPrivateKey?: string, lit?: any): Promise<ParsedResponseType[]> {
  const parsedResults = await Promise.all(response.map(async (apiNotification: ApiNotificationType) => {
    if (apiNotification.payload.data.secret) {
      apiNotification = await decryptFeed({ feed: apiNotification, pgpPrivateKey, lit });
    }
    const {
      payload: {
        data: {
          acta: cta = "",
          amsg: bigMessage = "",
          asub = "",
          icon = "",
          url = "",
          sid = "",
          app = "",
          aimg = "",
          secret = ""
        },
        notification,
      },
      source,
    } = apiNotification;
    return {
      cta,
      title: asub || '',
      message: bigMessage || notification.body || '',
      icon,
      url,
      sid,
      app,
      image: aimg,
      blockchain: source,
      notification,
      secret
    };
  }));

  return parsedResults;
}

File: packages/push-api-dart/lib/src/utils/parse_api.dart
import '../../push_api_dart.dart';

List<FeedResponse> parseApiResponse(List response) {
  return response.map((apiNotification) {
    final payload = apiNotification['payload'];
    final notification = apiNotification['payload']['notification'];

    return FeedResponse(
      cta: payload['cta'],
      title: payload['asub'] ?? '',
      message: payload['bigMessage'] ?? notification['body'] ?? '',
      icon: payload['icon'],
      url: payload['url'],
      sid: payload['sid'],
      app: payload['app'],
      image: payload['aimg'],
      blockchain: apiNotification['source'],
      secret: payload['secret'],
      notification: Notification.fromJson(notification),
    );
  }).toList();
}
